pkg/types/validation/installconfig.go

```diff
@@ -976,3 +978,52 @@ func validateAdditionalCABundlePolicy(c *types.InstallConfig) error {
 		return fmt.Errorf("supported values \"Proxyonly\", \"Always\"")
 	}
 }
+
+// validateFeatureSet returns an error if a gated feature is used without opting into the feature set.
+func validateFeatureSet(c *types.InstallConfig) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if _, ok := configv1.FeatureSets[c.FeatureSet]; !ok {
+		sortedFeatureSets := func() []string {
+			v := []string{}
+			for n := range configv1.FeatureSets {
+				v = append(v, string(n))
+			}
+			sort.Strings(v)
+			return v
+		}()
+		allErrs = append(allErrs, field.NotSupported(field.NewPath("featureSet"), c.FeatureSet, sortedFeatureSets))
+	}
+
+	if c.FeatureSet != configv1.TechPreviewNoUpgrade {
+		errMsg := "the TechPreviewNoUpgrade feature set must be enabled to use this field"
```

Comment:
Could CustomNoUpgrade also be allowed? Could there be a scenario where a custome feature and a tech preview need to be tested together?